metricbeat/module/couchdb/server/server.go

```diff
@@ -61,23 +71,108 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 	if err != nil {
 		return nil, err
 	}
+
 	return &MetricSet{
-		base,
-		http,
+		BaseMetricSet: base,
+		http:          http,
+		fetcher:       nil,
 	}, nil
 }
 
 // Fetch methods implements the data gathering and data conversion to the right
 // format. It publishes the event which is then forwarded to the output. In case
 // of an error set the Error field of mb.Event or simply call report.Error().
 func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
+	info, err := m.getInfoFromCouchdbHost(m.Host())
```

Comment:
Historically this was the preferred way to achieve this. Even in very heavy operations like opening connections to SQL databases. Frankly, I never liked very much the idea but I understood the implications (a long living SQL connection that may drop and leave the beat requiring a restart.

Revalidation using schema errors or any other more complex stuff is against the common "guides" (written nowhere) in Beats of trying to have explicit, easy to reason / easy to code / easy to maintain codebase. In Couchdb it's not a very good idea because the differences are very very small and can lead to unexpected results. An example are incomplete JSON returns by design or a mix of errors from the http transport layer with the database layer that can make your life a hell :smile: 

TLDR: I'll disagree and commit and move it to `New` :wink:

Reply:
Hmm... I wonder if it can lead to a potential problem - beats app starts before couchdb so it can't determine the API version.

Reply:
> Hmm... I wonder if it can lead to a potential problem - beats app starts before couchdb so it can't determine the API version.

Yes! We ran into this exact issue with a couple of the stack monitoring modules in Metricbeat and had to move all API calls into `Fetch`. See https://github.com/elastic/beats/issues/15258 (fixed by https://github.com/elastic/beats/pull/15270) and https://github.com/elastic/beats/issues/15276 (fixed by https://github.com/elastic/beats/pull/15306).

One solution I've seen in the past is caching such "infrequently changing" information for some period of time to save on checking it every `Fetch()` period. But that can introduce its own issues such as staleness so it might not be worth it. In general I err towards keeping it simple until it starts to be a problem.

Reply:
Regarding the offline discussion - I think you can move the code determining API version just before calling the right fetcher. Both fetchers should be stateless, so you freely initialize them in `New()` and get rid of any unnecessary mem allocations in `Fetch`.